File: 2021/Day19/Binary Search Tree to Greater Sum Tree/Binary Search Tree to Greater Sum Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    TreeNode *bstToGst(TreeNode *root)
    {
        int acc_sum = 0;
        sum_node(root, acc_sum);
        return root;
    }

    void sum_node(TreeNode *node, int &acc_sum)
    {
        if (node == NULL)
            return;
        sum_node(node->right, acc_sum);
        node->val += acc_sum;
        acc_sum = node->val;
        sum_node(node->left, acc_sum);
    }
};

File: 2021/Day19/Binary Search Tree to Greater Sum Tree/Binary Search Tree to Greater Sum Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:

        def recur(node, sumTillNow):
            if not node:
                return sumTillNow
            node.val += recur(node.right, sumTillNow)
            return recur(node.left, node.val)

        recur(root, 0)
        return root
